compiler/optimizer/VPHandlers.cpp

```diff
@@ -1151,31 +1151,92 @@ TR::Node *constrainBCDAggrLoad(OMR::ValuePropagation *vp, TR::Node *node)
-      if (index->getOpCode().isLoadConst() && array->getOpCode().isIndirect())
+      if (index->getOpCode().isLoadConst())
          {
          bool isGlobal;
-         TR::Node* base = array->getFirstChild();
-         TR::VPConstraint *baseVPConstraint = vp->getConstraint(base, isGlobal);
-         if (baseVPConstraint && baseVPConstraint->isConstString())
+         if (dataType == TR::Int16 && array->getOpCode().isIndirect())
             {
-            TR::VPConstString *constString = baseVPConstraint->getClassType()->asConstString();
+            TR::Node* base = array->getFirstChild();
+            TR::VPConstraint *baseVPConstraint = vp->getConstraint(base, isGlobal);
+            if (baseVPConstraint && baseVPConstraint->isConstString())
+               {
+               TR::VPConstString *constString = baseVPConstraint->getClassType()->asConstString();
 
-            uint16_t ch = constString->charAt(((TR::Compiler->target.is64Bit() ? index->getLongIntLow() : index->getInt())
-                                               - TR::Compiler->om.contiguousArrayHeaderSizeInBytes()) / 2, vp->comp());
-            if (ch != 0)
+               uint16_t ch = constString->charAt(((TR::Compiler->target.is64Bit() ? index->getLongIntLow() : index->getInt())
+                                                  - TR::Compiler->om.contiguousArrayHeaderSizeInBytes()) / 2, vp->comp());
+               if (ch != 0)
+                  {
+                  vp->replaceByConstant(node, TR::VPShortConst::create(vp, ch), isGlobal);
+                  return node;
+                  }
+               }
+            }
+         TR::VPConstraint *arrayVPConstraint = vp->getConstraint(array, isGlobal);
+         if (arrayVPConstraint)
+            {
+#ifdef J9_PROJECT_SPECIFIC
+            TR::KnownObjectTable *knot = vp->comp()->getKnownObjectTable();
+            TR::VPKnownObject *kobj = arrayVPConstraint->getKnownObject();
+            if (knot && kobj)
                {
-               vp->replaceByConstant(node, TR::VPIntConst::create(vp, ch), true);
-               return node;
+               TR::KnownObjectTable::Index idx = kobj->getIndex();
+               if (kobj->isArrayWithConstantElements(vp->comp()) && kobj->isPrimitiveArray(vp->comp()))
+                  {
+                  TR::VMAccessCriticalSection constrainAnyIntLoadCriticalSection(vp->comp(),
+                     TR::VMAccessCriticalSection::tryToAcquireVMAccess);
+                  if (constrainAnyIntLoadCriticalSection.hasVMAccess())
+                     {
+                     uintptrj_t arrayObj = knot->getPointer(idx);
+                     uintptrj_t offset = TR::Compiler->target.is64Bit()
+                                         ? index->getLongInt() : index->getInt();
+                     uintptrj_t lengthInElements = TR::Compiler->om.getArrayLengthInElements(vp->comp(), arrayObj);// * 2
+                     switch (kobj->getPrimitiveArrayDataType())
```

Comment:
The type of the access might not be the same as the array element type. Also, the bound checks for accesses larger than 1 byte won't detect very large offsets for which the addition overflows. Since the bound checks need logic that is slightly more involved, can we deduplicate them, e.g. something like this?
```
TR::DataType elementType = kobj->getPrimitiveArrayDataType();
uintptrj_t lengthInBytes = lengthInElements * TR::DataType::getSize(elementType);
uintptrj_t validStart = TR::Compiler->om.contiguousArrayHeaderSizeInBytes();
uintptrj_t validEnd = validStart + lengthInBytes;
uintptrj_t accessSize = TR::DataType::getSize(dataType);
uintptrj_t accessEnd = offset + accessSize;
if (validStart <= offset && offset < accessEnd && accessEnd <= validEnd)
   {
   switch (dataType)
      {
      //... read the memory and replaceByConstant()
      }
   }
```